File: nodejs/docker2go.js
const fs = require('fs')
const path = require('path')
const argv = require('yargs').argv

const args = argv._

let format = argv.format

function jsonInspectToCommand(fileContent) {
    const dockerConfig = JSON.parse(fileContent)[0]
    const fmtEnding = format ? ' \\\n' : ''

    let command = ''
    command += 'docker run -d'
    command += ' --name=' + dockerConfig.Name.substr(1)
    command += fmtEnding

    if (dockerConfig.HostConfig.AutoRemove) {
      command += " --rm";
    }

    if (dockerConfig.HostConfig.RestartPolicy.Name != 'no' && dockerConfig.HostConfig.RestartPolicy.Name != '') {
      command += ' --restart ' + dockerConfig.HostConfig.RestartPolicy.Name
      command += fmtEnding
    }

    let isHexa = /^[0-9a-fA-F]+$/.test(dockerConfig.Config.Hostname)
	  if (!isHexa) {
      command += " --hostname=" + dockerConfig.Config.Hostname
      command += fmtEnding
    }

    if (dockerConfig.HostConfig.Memory > 0) {
      // memory is stored in bytes, minumum is 4M
      command += ' -m '

      var kbMemory = parseInt(dockerConfig.HostConfig.Memory, 10) / 1024
      var mbMemory = kbMemory / 1024
      var gbMemory = mbMemory / 1024

      if (gbMemory >= 1 && Math.round(gbMemory) == gbMemory) {
        command += Math.round(gbMemory) + 'g'
          }
          else if (mbMemory >= 1 && Math.round(mbMemory) == mbMemory) {
        command += Math.round(mbMemory) + 'm'
          }
          else if (kbMemory >= 1 && Math.round(kbMemory) == kbMemory) {
        command += Math.round(kbMemory) + 'k'
          }
          else {
        command += dockerConfig.HostConfig.Memory + 'b'
      }
      command += fmtEnding
    }

    if (dockerConfig.HostConfig.NanoCpus > 0) {
      var cpus = dockerConfig.HostConfig.NanoCpus / 1000000000
		  command += " --cpus=" + cpus
    }

    if (dockerConfig.HostConfig.Dns != null) {
      for (let i=0; i < dockerConfig.HostConfig.Dns.length; i++) {
        command += ' --dns=' + dockerConfig.HostConfig.Dns[i]
        command += fmtEnding
      }
    }

    if (dockerConfig.Config.Env != null) {
      for (let i=0; i < dockerConfig.Config.Env.length; i++) {
              let env = dockerConfig.Config.Env[i]

        if (!(env.startsWith('PATH') || env.startsWith('LANG') || env.startsWith('LC_ALL'))) {
                  command += ' -e '

          if (env.indexOf(' ') != -1) {
            command += env.substr(0, env.indexOf('=')) + '="' + env.substr(env.indexOf('=')+1) + '"'
          } else {
            command += env
          }
          command += fmtEnding
        }
      }
    }

    if (dockerConfig.HostConfig.Binds != null) {
      for (let i=0; i < dockerConfig.HostConfig.Binds.length; i++) {
              let bind = dockerConfig.HostConfig.Binds[i]

        command += ' -v '
        if (bind.indexOf(' ') != -1) {
          bind = '"' + bind + '"'
        }
        command += bind
        command += fmtEnding
      }
    }

    if (dockerConfig.HostConfig.PortBindings != null) {
      for (let port in dockerConfig.HostConfig.PortBindings) {
              let binding = dockerConfig.HostConfig.PortBindings[port]

        for (let i=0; i < binding.length; i++) {
          command += ' -p ' + binding[i].HostPort + ':' + port.split('/')[0]
          command += fmtEnding
        }
      }
    }

    if (dockerConfig.HostConfig.Links != null) {
      for (let i=0; i < dockerConfig.HostConfig.Links.length; i++) {
              let splitted = dockerConfig.HostConfig.Links[i].split('/')

        let preLink = splitted[1].substr(0, splitted[1].length-1)
              command += ' --link ' + preLink

        if (preLink != splitted[3]) {
          command += ':' + splitted[3]
        }

        command += fmtEnding
      }
    }

    if (Object.keys(dockerConfig.HostConfig.LogConfig.Config).length > 0) {
      for (let config in dockerConfig.HostConfig.LogConfig.Config) {
        command += ' --log-opt ' + config + '=' + dockerConfig.HostConfig.LogConfig.Config[config]
      }
      command += fmtEnding
    }

    command += ' ' + dockerConfig.Config.Image

    return command
}

if (args.length > 0) {
    for (let i = 0; i < args.length; i++) {
        const fileName = path.resolve(args[i])

        if (!fs.statSync(fileName).isDirectory()) {
            const fileContent = fs.readFileSync(fileName, 'utf8')

            const command = jsonInspectToCommand(fileContent)

            if (args.length > 1) {
                console.log("File:", path.basename(fileName))
                console.log(command)

                if (i < (args.length - 1)) {
                    console.log()
                }
            } else {
                console.log(command)
            }
        }
    }
}
else if (!process.stdin.isTTY) {
    process.stdin.on( 'data', function(data) {
        let command = jsonInspectToCommand(data)
        console.log(command)
    });
}
else {
    console.log('Usage:')
    console.log('./docker2go docker_inspect_of_a_container.txt OR')
    console.log('docker inspect <container> | docker2go')
}

File: go/docker2go.go
package main

import (
	"bufio"
	"encoding/hex"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"math"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

type DockerConfig struct {
	Name   string
	Config struct {
		Env      []string
		Hostname string
		Image    string
	}
	HostConfig struct {
		AutoRemove bool
		Binds      []string
		Dns        []string
		Links      []string
		LogConfig  struct {
			Config map[string]string
		}
		Memory       int
		NanoCpus     int
		PortBindings map[string][]struct {
			HostIp   string
			HostPort string
		}
		RestartPolicy struct {
			Name string
		}
	}
}

var dockerConfigs []DockerConfig
var format = false

func jsonInspectToCommand(data []byte) (string, error) {
	dockerConfigs = nil
	var fmtEnding = ""
	if format {
		fmtEnding = " \\\n"
	}

	err := json.Unmarshal(data, &dockerConfigs)
	if err != nil {
		fmt.Println("Error parsing json:", err)
		return "", err
	}

	var dockerConfig = dockerConfigs[0]

	var command string
	command += "docker run -d"
	command += " --name=" + dockerConfig.Name[1:]
	command += fmtEnding

	if dockerConfig.HostConfig.AutoRemove {
		command += " --rm"
	}

	if dockerConfig.HostConfig.RestartPolicy.Name != "no" && dockerConfig.HostConfig.RestartPolicy.Name != "" {
		command += " --restart " + dockerConfig.HostConfig.RestartPolicy.Name
		command += fmtEnding
	}

	_, err = hex.DecodeString(dockerConfig.Config.Hostname)
	if err != nil {
		command += " --hostname=" + dockerConfig.Config.Hostname
		command += fmtEnding
	}

	if dockerConfig.HostConfig.Memory > 0 {
		// memory is stored in bytes, minumum is 4M
		command += " -m "

		var kbMemory = float64(dockerConfig.HostConfig.Memory) / 1024
		var mbMemory = kbMemory / 1024
		var gbMemory = mbMemory / 1024

		if gbMemory >= 1 && math.Round(gbMemory) == gbMemory {
			command += strconv.Itoa(int(gbMemory)) + "g"
		} else if mbMemory >= 1 && math.Round(mbMemory) == mbMemory {
			command += strconv.Itoa(int(mbMemory)) + "m"
		} else if kbMemory >= 1 && math.Round(kbMemory) == kbMemory {
			command += strconv.Itoa(int(kbMemory)) + "k"
		} else {
			command += strconv.Itoa(dockerConfig.HostConfig.Memory) + "b"
		}
		command += fmtEnding
	}

	if dockerConfig.HostConfig.NanoCpus > 0 {
		var cpus = float64(dockerConfig.HostConfig.NanoCpus) / 1000000000
		command += " --cpus=" + strconv.FormatFloat(cpus, 'f', -1, 64)
	}

	if len(dockerConfig.HostConfig.Dns) > 0 {
		for _, dns := range dockerConfig.HostConfig.Dns {
			command += " --dns=" + dns
		}
		command += fmtEnding
	}

	if dockerConfig.Config.Env != nil {
		for _, env := range dockerConfig.Config.Env {
			if !(strings.HasPrefix(env, "PATH") || strings.HasPrefix(env, "LANG") || strings.HasPrefix(env, "LC_ALL")) {
				command += " -e "
				if strings.IndexAny(env, " &;") != -1 {
					command += env[0:strings.Index(env, "=")] + "=\"" + env[strings.Index(env, "=")+1:] + "\""
				} else {
					command += env
				}
				command += fmtEnding
			}
		}
	}

	if dockerConfig.HostConfig.Binds != nil {
		for _, bind := range dockerConfig.HostConfig.Binds {
			command += " -v "
			if strings.IndexAny(bind, " &;") != -1 {
				bind = "\"" + bind + "\""
			}
			command += bind
			command += fmtEnding
		}
	}

	if dockerConfig.HostConfig.PortBindings != nil {
		for port, bindigs := range dockerConfig.HostConfig.PortBindings {
			for _, binding := range bindigs {
				command += " -p " + binding.HostPort + ":" + strings.Split(port, "/")[0]
				command += fmtEnding
			}
		}
	}

	if dockerConfig.HostConfig.Links != nil {
		for _, link := range dockerConfig.HostConfig.Links {
			splitted := strings.Split(link, "/")
			preLink := splitted[1][:len(splitted[1])-1]
			command += " --link " + preLink
			if preLink != splitted[3] {
				command += ":" + splitted[3]
			}
			command += fmtEnding
		}
	}

	if len(dockerConfig.HostConfig.LogConfig.Config) > 0 {
		for configName, configValue := range dockerConfig.HostConfig.LogConfig.Config {
			command += " --log-opt " + configName + "=" + configValue
		}
		command += fmtEnding
	}

	command += " " + dockerConfig.Config.Image
	return command, nil
}

func main() {
	var data []byte

	fmtFlag := flag.Bool("format", false, "format the output")

	flag.Parse()

	format = *fmtFlag
	args := flag.Args()

	if len(args) > 0 {
		for i := 0; i < len(args); i++ {
			fileName, err := filepath.Abs(args[i])

			fileInfo, err := os.Stat(fileName)
			if err != nil {
				fmt.Println("Error checking file", fileName, err)
				continue
			}

			if !fileInfo.IsDir() {
				fileContent, err := ioutil.ReadFile(fileName)
				if err != nil {
					fmt.Println("Error reading file", fileName, err)
					continue
				}

				command, err := jsonInspectToCommand(fileContent)
				if err != nil {
					fmt.Println("Error obtaining command from file", fileName, err)
					continue
				}

				if len(args) > 1 {
					fmt.Println("File:", filepath.Base(fileName))
					fmt.Println(command)
					if i < len(args)-1 {
						fmt.Println()
					}
				} else {
					fmt.Println(command)
				}
			}
		}
		os.Exit(0)
	} else {
		scanner := bufio.NewScanner(os.Stdin)
		var sdtinData string
		for scanner.Scan() {
			sdtinData += scanner.Text()
		}

		if scanner.Err() != nil {
			fmt.Println("Error reading stdin", scanner.Err())
			os.Exit(1)
		}
		data = []byte(sdtinData)

		command, err := jsonInspectToCommand(data)
		if err != nil {
			fmt.Println("Error obtaining command from json", err)
			os.Exit(1)
		}

		fmt.Println(command)
	}

	if len(data) == 0 {
		fmt.Println("Usage:")
		fmt.Println("./docker2go docker_inspect_of_a_container.txt OR")
		fmt.Println("docker inspect <container> | docker2go")
		os.Exit(0)
	}
}
